Maple2.Server.Game/PacketHandlers/MasteryHandler.cs

```diff
@@ -0,0 +1,111 @@
+    }
+
+    private void HandleCraft(GameSession session, IByteReader packet) {
+        int recipeId = packet.ReadInt();
+
+        if (!TableMetadata.MasteryRecipeTable.Entries.TryGetValue(recipeId, out MasteryRecipeTable.Entry? entry)) {
+            return;
+        }
+
+        // TODO: Check if player has completed the required quests.
+
+        if (session.Mastery[entry.Type] < entry.RequiredMastery) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_mastery));
+        }
+
+        if (session.Currency.CanAddMeso(-entry.RequiredMeso) != -entry.RequiredMeso) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_meso));
+            return;
+        }
+
+        Dictionary<int, IEnumerable<Item>> itemDictionary = new Dictionary<int, IEnumerable<Item>>();
+        foreach (MasteryRecipeTable.Ingredient ingredient in entry.RequiredItems) {
+            IEnumerable<Item>? foundItems = session.Item.Inventory.Find(ingredient.ItemId, ingredient.Rarity).ToList();
+            int totalCount = foundItems.Sum(item => item.Amount);
+            if (totalCount < ingredient.Amount) {
+                session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_item));
+                return;
+            }
+            itemDictionary.Add(ingredient.ItemId, foundItems);
+        }
+
+        if (entry.RequiredItems.Any(ingredient => session.Item.Inventory.Find(ingredient.ItemId, ingredient.Rarity).ToList().Count < ingredient.Amount)) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_item));
+            return;
+        }
+
+        session.Currency.Meso -= entry.RequiredMeso;
+
+        if (!entry.NoRewardExp) {
+            session.Mastery[entry.Type] += entry.RewardMastery;
+            session.Send(MasteryPacket.UpdateMastery(entry.Type, session.Mastery[entry.Type]));
+        }
+
+        foreach (MasteryRecipeTable.Ingredient ingredient in entry.RequiredItems) {
+            IEnumerable<Item> itemList = itemDictionary[ingredient.ItemId];
+            int amountToConsume = ingredient.Amount;
+            foreach (Item item in itemList) {
+                if (amountToConsume == 0) {
+                    break;
+                }
+                if (item.Amount >= amountToConsume) {
+                    session.Item.Inventory.Consume(item.Uid, amountToConsume);
+                    break;
+                }
+
+                session.Item.Inventory.Consume(item.Uid, item.Amount);
+                amountToConsume -= item.Amount;
+            }
+        }
+
+        foreach (MasteryRecipeTable.Ingredient rewardItem in entry.RewardItems) {
+            if (!ItemMetadata.TryGet(rewardItem.ItemId, out ItemMetadata? itemMetadata)) {
+                continue;
+            }
+            Item item = new(itemMetadata, rewardItem.Rarity, rewardItem.Amount);
+            if (!session.Item.Inventory.CanAdd(item)) {
+                // TODO: Mail to player
```

Comment:
What I want instead of some kind of "SafeAdd" on inventory, so that items that cannot be added will automatically be sent to mail. I think there are lots of cases where this is needed.

Reply:
No action here, just FYI.